nautobot/core/api/renderers.py

```diff
@@ -24,3 +36,168 @@ class NautobotJSONRenderer(JSONRenderer):
+                base_headers.remove(undesired_header)
+
+        # Remove all write-only and object-count headers
+        for header in list(base_headers):
+            if serializer.fields[header].write_only:
+                base_headers.remove(header)
+            if header.endswith("_count"):
+                base_headers.remove(header)
+
+        # Add individual headers for each relevant custom field
+        if "custom_fields" in serializer.fields:
+            cf_keys = CustomField.objects.get_for_model(serializer.Meta.model).values_list("key", flat=True)
+            cf_headers = sorted([f"cf_{key}" for key in cf_keys])
+        else:
+            cf_headers = []
+
+        # TODO: relationships? computed fields?
+
+        headers = base_headers + cf_headers
+
+        logger.debug("CSV headers for %s are: %s", type(serializer).__name__, headers)
+        return headers
+
+    def object_to_row_elements(self, obj, *, headers, serializer):
+        """Given an object, the desired CSV headers, and its serializer, yield the serialized values for each header."""
+        for key in headers:
+            serializer_field = serializer.fields.get(key, None)
+
+            # Retrieve the base value corresponding to this key
+            if key.startswith("cf_"):
+                # Custom field
+                value = obj._custom_field_data.get(key[3:], None)
+            elif isinstance(serializer_field, serializers.ListSerializer):
+                # A list of related objects;  as CSV doesn't really have a good way to render a list of lists into
+                # a single field, we use the natural-key-slugs for each object instead of the natural keys
+                related_queryset = serializer_field.get_attribute(obj)
+                value = [
+                    construct_natural_key_slug(self.get_natural_key(related_object))
+                    for related_object in related_queryset.all()
+                ]
+            elif isinstance(serializer_field, serializers.Serializer):
+                # A related object; we want to render its natural key if at all possible.
+                related_object = serializer_field.get_attribute(obj)
+                value = self.get_natural_key(related_object)
+            elif isinstance(serializer_field, serializers.SerializerMethodField):
+                # Tricky since a SerializerMethodField can theoretically return anything.
+                value = serializer.data[key]
+                # However, in *many* cases in Nautobot core, it's just a wrapper around a GenericForeignKey field.
+                # So let's try that:
+                try:
+                    related_object = getattr(obj, key)
+                    if related_object is None:
+                        value = None
+                    elif isinstance(related_object, models.Model):
+                        value = [related_object._meta.label_lower] + self.get_natural_key(related_object)
```

Comment:
Why do we need to wrap a `[]` around `related_object._meta.label_lower`?

Reply:
The natural key is a list of values, so this is prepending the content-type to that list. I may need to adjust this as part of #3715 as I'm finding that mixing "natural keys as a comma-separated list" and "natural keys as a natural-key-slug" is challenging to handle consistently on the import side.